### PR TITLE
Add timelineEvents to DBPUIDataBrokerProfileMatch

### PR DESCRIPTION
**Pull request by quanganhdo**
Task/Issue URL: https://app.asana.com/0/0/1208663928051298/f
Tech Design URL: https://app.asana.com/0/481882893211075/1208663928051302
CC:

**Description**:

Replace existing date fields with timelineEvents

**Steps to test this PR**:
1. CI passes

**Definition of Done**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

---
###### Internal references:
[Pull Request Review Checklist](https://app.asana.com/0/1202500774821704/1203764234894239/f)
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
[Pull Request Documentation](https://app.asana.com/0/1202500774821704/1204012835277482/f)

### CODE REVIEW

File: LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift [hunk 1]
```diff
@@ -301,26 +367,20 @@ struct DBPUIOptOutMatch: DBPUISendableMessage {
...
-    let optOutSubmittedDate: Double?
-    let estimatedRemovalDate: Double?
-    let removedDate: Double?
+    let timelineEvents: [DBPUITimelineEvent]
 }
 
 extension DBPUIOptOutMatch {
     init?(profileMatch: DBPUIDataBrokerProfileMatch, matches: Int) {
-        guard let removedDate = profileMatch.removedDate else { return nil }
+        guard let removedDate = profileMatch.timelineEvents.filter({ $0.type == .removed }).first?.date else { return nil }
```
Comment (quanganhdo) [inline LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift:375]:
```suggestion
        guard let removedDate = profileMatch.timelineEvents.first(where: { $0.type == .removed })?.date else { return nil }
```

File: LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift [hunk 2]
```diff
@@ -146,11 +146,93 @@ struct DBPUIDataBrokerList: DBPUISendableMessage {
...
+        let mostRecentFoundEvent = timelineEvents.filter({ $0.type == .recordFound || $0.type == .recordReappeared }).max(by: <)
+        let mostRecentOptOutSubmittedEvent = timelineEvents.filter({ $0.type == .optOutSubmitted }).max(by: <)
+
+        if let optOutSubmittedDate = (mostRecentOptOutSubmittedEvent ?? mostRecentFoundEvent)?.eventDate,
+           let estimatedRemovalEvent = DBPUITimelineEvent(estimatedRemovalDate: Calendar.current.date(byAdding: .day, value: 14, to: optOutSubmittedDate)) {
+            timelineEvents.append(estimatedRemovalEvent)
+        }
+
+        if let removedEvent = DBPUITimelineEvent(removedDate: removedDate) {
+            timelineEvents.append(removedEvent)
```
Comment (THISISDINOSAUR) [inline LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift:223]: Am I right in thinking that if the entry has been removed, then the timeline will have an estimated removal date event, followed by an actual removed event? Is that what we want?

File: LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift [hunk 3]
```diff
@@ -146,11 +146,93 @@ struct DBPUIDataBrokerList: DBPUISendableMessage {
...
+        case recordReappeared = "record-reappeared"
+        case optOutSubmitted = "opt-out-submitted"
+        case estimatedRemoval = "estimated-removal"
+        case removed = "removed"
+    }
+
+    let type: EventType
+    let date: Double
+
+    var eventDate: Date {
```
Comment (THISISDINOSAUR) [inline LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift:168]: I think eventDate has a slightly confusing name. How about dateAsSwiftDate?

**Review by THISISDINOSAUR — COMMENTED** [review]
Have you been able to test with FE at all? Is there any impact on performance?

File: LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift (hunk 2 above)
Comment (quanganhdo) [reply LocalPackages/DataBrokerProtection/Sources/DataBrokerProtection/Model/DBPUICommunicationModel.swift:223]: Ah good point. Logically we shouldn't have the estimated removal date when the profile has been removed 😅

**Comment by quanganhdo** [comment]
Closing this PR, let's revisit it later outside the scope of my project.

**Pull request closed without merging**